app/app.go

```diff
@@ -477,40 +510,79 @@ func (app *CyberdApp) EndBlocker(ctx sdk.Context, _ abci.RequestEndBlock) abci.R
 	app.currentCreditPrice = newPrice
 	app.curBlockSpentBandwidth = 0
 
-	linksCount := app.mainKeeper.GetLinksCount(ctx)
-	cidsCount := app.mainKeeper.GetCidsCount(ctx)
+	// START RANK CALCULATION
+	//linksCount := app.mainKeeper.GetLinksCount(ctx)
 
-	start := time.Now()
-	calcCtx := rank.NewCalcContext(ctx, app.linkIndexedKeeper, app.cidNumKeeper, *app.stakeIndex)
-	newRank, steps := rank.CalculateRank(calcCtx, app.computeUnit, app.BaseApp.Logger)
-	app.BaseApp.Logger.Info(
-		"Rank calculated", "steps", steps, "time", time.Since(start), "links", linksCount, "cids", cidsCount,
-	)
+	//start := time.Now()
+	if ctx.BlockHeight()%rank.CalculationPeriod == 0 || ctx.BlockHeight() == 1 {
 
-	app.latestBlockHeight = ctx.BlockHeight()
-	app.mainKeeper.StoreLatestBlockNumber(ctx, uint64(ctx.BlockHeight()))
+		if !app.rankCalculationFinished {
+			select {
+			case newRank := <-app.rankCalcChan:
+				app.handleNewRank(ctx, newRank)
+			case err := <-app.rankErrChan:
+				// DUMB ERROR HANDLING
+				app.BaseApp.Logger.Error("Error during rank calculation " + err.Error())
+				panic(err.Error())
+			}
+		}
+
+		app.rankState.ApplyNextRank(app.rankCidCount)
+		// Recalculate index
+		go app.rankState.BuildCidRankedLinksIndexInParallel(app.rankCidCount)
```

Comment:
A bit later we fix links. Is it concurrent modification for `BuildCidRankedLinksIndexInParallel` ?